File: plugins/ios-widget/widget/WidgetSnapshot.swift
//
//  WidgetSnapshot.swift
//  UptimePocketWidget
//
//  The data shape the widget reads from disk. Mirrors
//  src/platform/widget/snapshot.ts (the app-side source of truth).
//
//  Why Codable? The app writes JSON, the widget reads it. We avoid
//  a hand-rolled parser; if the JSON doesn't decode, the widget
//  shows a "no data" placeholder rather than crashing.
//

import Foundation

/// Status of a single monitor. Mirrors `MonitorStatus` in
/// `src/domain/models`. The order matters for "worst status"
/// ranking — the widget shows the most-actionable item first.
enum WidgetMonitorStatus: String, Codable {
    case up
    case down
    case pending
    case maintenance
    case paused
}

/// A single monitor row. Field names match the TypeScript
/// source (camelCase) exactly so the JSON decode is symmetric.
struct WidgetMonitor: Codable, Identifiable {
    /// Stable ID. The app prefixes server ID so two servers
    /// can't collide on monitor ID.
    let id: String
    let name: String
    let status: WidgetMonitorStatus
    /// Unix ms. The widget renders "5m ago" for stale snapshots.
    let lastCheckAt: Int64?
    let responseTime: Int64?
    let serverLabel: String
}

/// A server: connection state + a flat list of monitors.
/// The widget iterates `monitors` directly (no nested traversal).
struct WidgetServer: Codable, Identifiable {
    let id: String
    let name: String
    let connected: Bool
    let worstStatus: WidgetMonitorStatus
    let monitors: [WidgetMonitor]
}

/// Top-level snapshot. Versioned so a future incompatible
/// change can be detected and the widget can fall back to
/// "no data" instead of crashing.
struct WidgetSnapshot: Codable {
    let generatedAt: Int64
    let version: Int
    let servers: [WidgetServer]

    /// Decode with a fallback to nil on any error. The widget
    /// shows "no data" when this returns nil.
    static func decode(from data: Data) -> WidgetSnapshot? {
        let decoder = JSONDecoder()
        do {
            let snap = try decoder.decode(WidgetSnapshot.self, from: data)
            // Future-proof: reject unknown versions.
            if snap.version != 1 { return nil }
            return snap
        } catch {
            return nil
        }
    }

    /// True if this snapshot is "fresh" (within the last 10 min).
    /// Used to show a "stale" badge on the widget if the app
    /// has been background-killed for a long time.
    func isFresh(nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> Bool {
        let ageMs = nowMs - generatedAt
        return ageMs >= 0 && ageMs < 10 * 60 * 1000
    }

    /// All monitors across all servers, sorted: down first,
    /// then by lastCheckAt desc. The widget shows the top N
    /// from this list.
    func prioritizedMonitors() -> [WidgetMonitor] {
        var all: [WidgetMonitor] = []
        for server in servers {
            all.append(contentsOf: server.monitors)
        }
        // Ranking mirrors TS: down > pending > maintenance > paused > up.
        let rank: [WidgetMonitorStatus: Int] = [
            .down: 5,
            .pending: 4,
            .maintenance: 3,
            .paused: 2,
            .up: 1,
        ]
        return all.sorted { a, b in
            let ra = rank[a.status] ?? 0
            let rb = rank[b.status] ?? 0
            if ra != rb { return ra > rb }
            return (a.lastCheckAt ?? 0) > (b.lastCheckAt ?? 0)
        }
    }
}

File: plugins/android-widget/android/app/src/main/java/de/quavon/uptimepocket/widget/WidgetSnapshot.kt
package de.quavon.uptimepocket.widget

/**
 * Snapshot of the app state, written by the React app to
 * `filesDir/widget_snapshot.json` and read by the Glance widget.
 *
 * The shape mirrors the TypeScript `WidgetSnapshot` in
 * `src/platform/widget/snapshot.ts`. Keep them in sync.
 *
 * Why a separate data class instead of reusing the React types?
 * The widget process can't share a runtime with the main app.
 * It has its own classloader and its own set of dependencies.
 * A standalone data class means no shared schema code and
 * no risk of an Android-only schema change breaking iOS.
 */
data class WidgetSnapshot(
    val generatedAt: Long,
    val version: Int,
    val servers: List<WidgetServer>
)

data class WidgetServer(
    val id: String,
    val name: String,
    val connected: Boolean,
    /** The worst status across the server's monitors. */
    val worstStatus: String,
    val monitors: List<WidgetMonitor>
)

data class WidgetMonitor(
    /** Composite id: `${serverId}::${monitorId}`. */
    val id: String,
    val name: String,
    val status: String,
    val lastCheckAt: Long?,
    val responseTime: Int?,
    val serverLabel: String
)

/** Status enum mirrored from TypeScript. Values must match. */
enum class MonitorStatus(val wire: String) {
    UP("up"),
    DOWN("down"),
    PENDING("pending"),
    MAINTENANCE("maintenance"),
    PAUSED("paused");

    companion object {
        fun fromWire(s: String?): MonitorStatus = when (s) {
            "up" -> UP
            "down" -> DOWN
            "pending" -> PENDING
            "maintenance" -> MAINTENANCE
            "paused" -> PAUSED
            else -> PENDING // unknown → safe default
        }
    }
}
